// @internet-of-people/sdk/dist/layer1/signed/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const operation_type_1 = require("../operation-type");
const visitor_1 = require("../visitor");
const add_key_1 = require("./add-key");
const revoke_key_1 = require("./revoke-key");
const add_right_1 = require("./add-right");
const revoke_right_1 = require("./revoke-right");
const tombstone_did_1 = require("./tombstone-did");
class SignableSchemaVisitor {
    addKey() {
        return add_key_1.addKeySchema();
    }
    revokeKey() {
        return revoke_key_1.revokeKeySchema();
    }
    addRight() {
        return add_right_1.addRightSchema();
    }
    revokeRight() {
        return revoke_right_1.revokeRightSchema();
    }
    tombstoneDid() {
        return tombstone_did_1.tombstoneDidSchema();
    }
}
const signableOperationSchemas = () => {
    return visitor_1.visitAllSignableOperationTypes(new SignableSchemaVisitor());
};
exports.signedSchema = () => {
    return {
        required: ['operation', 'signerPublicKey', 'signature', 'signables'],
        // additionalProperties: false, // TODO: https://github.com/ArkEcosystem/core/issues/3340
        properties: {
            operation: {
                type: 'string',
                const: operation_type_1.OperationType.Signed,
            },
            signables: {
                type: 'array',
                items: {
                    anyOf: signableOperationSchemas(),
                },
            },
            signerPublicKey: {
                type: 'string',
            },
            signature: {
                type: 'string',
            },
        },
    };
};
