// @internet-of-people/sdk/dist/layer1/signed/to-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const operation_type_1 = require("../operation-type");
class ToSignableDataVisitor {
    addKey(did, lastTxId, auth, expiresAtHeight) {
        const result = {
            operation: operation_type_1.SignableOperationType.AddKey,
            did: did.toString(),
            lastTxId,
            auth: auth.toString(),
        };
        if (expiresAtHeight) {
            result.expiresAtHeight = expiresAtHeight;
        }
        return result;
    }
    revokeKey(did, lastTxId, auth) {
        const result = {
            operation: operation_type_1.SignableOperationType.RevokeKey,
            did: did.toString(),
            lastTxId,
            auth: auth.toString(),
        };
        return result;
    }
    addRight(did, lastTxId, auth, right) {
        const result = {
            operation: operation_type_1.SignableOperationType.AddRight,
            did: did.toString(),
            lastTxId,
            auth: auth.toString(),
            right,
        };
        return result;
    }
    revokeRight(did, lastTxId, auth, right) {
        const result = {
            operation: operation_type_1.SignableOperationType.RevokeRight,
            did: did.toString(),
            lastTxId,
            auth: auth.toString(),
            right,
        };
        return result;
    }
    tombstoneDid(did, lastTxId) {
        const result = {
            operation: operation_type_1.SignableOperationType.TombstoneDid,
            did: did.toString(),
            lastTxId,
        };
        return result;
    }
}
exports.toSignableData = (operation) => {
    return operation.accept(new ToSignableDataVisitor());
};
